pkg/controllers/management/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -441,55 +497,83 @@ func (m *MCAppController) deleteApps(mcAppName string, mcapp *v3.MultiClusterApp
 				return nil, err
 			}
 			appsToDelete = append(appsToDelete, apps...)
+			projectToCluster[projectNS] = split[0]
 		}
 	}
-
-	if err := m.delete(appsToDelete); err != nil {
+	if err := m.delete(appsToDelete, clusterContexts, projectToCluster); err != nil {
 		return nil, err
 	}
 	return nil, nil
 }
 
-func (m *MCAppController) getAllApps(mcAppName string, all bool) ([]*pv3.App, error) {
+func (m *MCAppManager) getAllAppsWithContexts(mcAppName string) ([]*pv3.App, map[string]string, map[string]*config.UserContext, error) {
 	// to get all apps, get all clusters first, then get all apps in all projects of all clusters
 	appsToDelete := []*pv3.App{}
-	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcAppName})
+	projectToCluster := map[string]string{}
+	clusterContexts := map[string]*config.UserContext{}
+	set := labels.Set(map[string]string{MultiClusterAppIDSelector: mcAppName})
 	clusters, err := m.clusterLister.List("", labels.NewSelector())
 	if err != nil {
-		return appsToDelete, err
+		return appsToDelete, projectToCluster, clusterContexts, err
 	}
 	for _, c := range clusters {
-		if !all && m.clusterName != c.Name {
-			continue
+		projects, err := m.projectLister.List(c.Name, labels.NewSelector())
+		if err != nil {
+			return appsToDelete, projectToCluster, clusterContexts, err
+		}
+		for _, p := range projects {
+			apps, err := m.appLister.List(p.Name, set.AsSelector())
+			if err != nil {
+				return appsToDelete, projectToCluster, clusterContexts, err
+			}
+			appsToDelete = append(appsToDelete, apps...)
+			projectToCluster[p.Name] = c.Name
+		}
+		userContext, err := m.clusterManager.UserContext(c.Name)
+		if err != nil {
+			return appsToDelete, projectToCluster, clusterContexts, err
 		}
+		clusterContexts[c.Name] = userContext
+	}
+	return appsToDelete, projectToCluster, clusterContexts, err
+}
+
+func (m *MCAppManager) getAllApps(mcAppName string) ([]*pv3.App, map[string]string, error) {
```

Comment:
this method looks similar to getAllAppsWithContexts, lets see if we can either switch to using just one, or re-use the common parts (iterating over clusters/projects)

Reply:
thanks, the only difference was if clusterContexts was returned, updated definition to include `withContext` bool